File: cpp/leetcode/leetcode/12_integer_to_roman.cpp
//
//  12_integer_to_roman.cpp
//  leetcode
//
//  Created by hatlonely on 16/1/11.
//  Copyright © 2016年 hatlonely. All rights reserved.
//

#include <iostream>
#include <cassert>
#include <string>

namespace integer_to_roman {
    
class Solution {
public:
    std::string intToRoman(int num) {
        static const char* table[3][10] = {
            {"", "I", "II", "III", "IV", "V", "VI", "VII", "VIII", "IX"},
            {"", "X", "XX", "XXX", "XL", "L", "LX", "LXX", "LXXX", "XC"},
            {"", "C", "CC", "CCC", "CD", "D", "DC", "DCC", "DCCC", "CM"},
        };
        
        int n0 = num % 10;
        num /= 10;
        int n1 = num % 10;
        num /= 10;
        int n2 = num % 10;
        num /= 10;
        
        std::string roman;
        for (int i = 0; i < num; i++) {
            roman += "M";
        }
        
        roman += table[2][n2];
        roman += table[1][n1];
        roman += table[0][n0];
        
        return roman;
    }
};

int main(int argc, const char *argv[]) {
    Solution solution;
    
    assert(solution.intToRoman(25) == "XXV");
    
    return 0;
}
    
}

File: go/src/main/12_integer_to_roman.go
package main

import "fmt"

var kIntRomanMap = map[int]string{
	1: "I", 2: "II", 3: "III", 4: "IV", 5: "V", 6: "VI", 7: "VII", 8: "VIII", 9: "IX",
	10: "X", 20: "XX", 30: "XXX", 40: "XL", 50: "L", 60: "LX", 70: "LXX", 80: "LXXX", 90: "XC",
	100: "C", 200: "CC", 300: "CCC", 400: "CD", 500: "D", 600: "DC", 700: "DCC", 800: "DCCC", 900: "CM",
	1000: "M", 2000: "MM", 3000: "MMM",
	0: "",
}

func intToRoman(num int) (roman string) {
	return kIntRomanMap[num/1000*1000] + kIntRomanMap[(num%1000)/100*100] + kIntRomanMap[(num%100)/10*10] + kIntRomanMap[num%10]
}

func main() {
	for _, unit := range []struct {
		num   int
		roman string
	}{
		{100, "C"},
		{1980, "MCMLXXX"},
	} {
		if roman := intToRoman(unit.num); roman != unit.roman {
			fmt.Printf("intToRoman(%v) = %v != %v\n", unit.num, roman, unit.roman)
		}
	}
}
